client/txn.go

```diff
@@ -300,18 +301,33 @@ func (txn *Txn) CommitInBatch(b *Batch) error {
 
 // Commit sends an EndTransactionRequest with Commit=true.
 func (txn *Txn) Commit() error {
-	if txn.Proto.Writing {
-		return txn.sendEndTxnCall(true /* commit */)
+	err := txn.sendEndTxnCall(true /* commit */)
+	if err == nil {
+		return nil
 	}
-	return nil
+	if !txn.execing {
+		// If this transaction is not part of the auto-retry loop we need to
+		// cleanup the transaction on retryable commit failures.
+		if restartErr, ok := err.(proto.TransactionRestartError); ok {
+			if restartErr.CanRestartTransaction() != proto.TransactionRestart_ABORT {
+				if replyErr := txn.Rollback(); replyErr != nil {
+					// Log the error only if it is not a TransactionAbortedError.
+					if _, ok := replyErr.(*proto.TransactionAbortedError); !ok {
+						log.Errorf("failure aborting transaction: %s; abort caused by: %s", replyErr, err)
+					}
+				}
+			}
+		}
+	}
+	return err
 }
 
 // Rollback sends an EndTransactionRequest with Commit=false.
 func (txn *Txn) Rollback() error {
-	var err error
-	if txn.Proto.Writing {
-		err = txn.sendEndTxnCall(false /* commit */)
```

Comment:
if we're not writing, do we still need to send the EndTransaction call? I don't understand this change.